lib/pbench/server/sync.py

```diff
@@ -0,0 +1,138 @@
+
+        NOTE:
+
+        This can be considered a prototype for a general "query by metadata"
+        mechanism (PBENCH-825), however I haven't attempted to work out a full
+        generalization. At this point I'm happen enough to have a relatively
+        simple special purpose query; but it bring hope that the general query
+        is achievable.
+
+        Args:
+            A desired Operation enum value
+
+        Returns:
+            A list of Dataset objects which have an associated
+            "server.operation" metadata value containing the name of the
+            operation enum.
+        """
+        try:
+            query = Database.db_session.query(Dataset).join(Metadata)
+            query = (
+                query.filter(Metadata.key == Metadata.SERVER)
+                .filter(
+                    Metadata.value["operation"].as_string().contains(operation.name)
+                )
+                .order_by(Dataset.name)
+            )
+            if self.logger.isEnabledFor(DEBUG):
+                self.logger.debug(
+                    "QUERY {}",
+                    query.statement.compile(compile_kwargs={"literal_binds": True}),
+                )
+            return list(query.all())
+        except Exception as e:
+            self.logger.exception("Failed to query for {}", operation)
+            raise SyncSqlError("next") from e
+
+    def update(
+        self,
+        dataset: Dataset,
+        did: Optional[Operation] = None,
+        enabled: Optional[List[Operation]] = None,
+        status: Optional[str] = None,
+    ):
+        """
+        Advertise the operations for which the dataset is now ready.
+
+        TODO: It'd be nice to nest this inside a transaction so it's all
+        atomic. This seems to be difficult to achieve in a manner that works
+        both with PostgreSQL for production and sqlite3 for unit tests. When
+        we've removed the use of sqlite3 we can make this a nested transaction.
+
+        Args:
+            dataset: The dataset
+            did: The operation (if any) just completed, which will marked done
+            enabled: A list (if any) of operations for which the dataset is now
+                eligible.
+            status: A status message (if not specified, and enabling new
+                operation(s), the default is "ok")
+        """
+        operations: List[str] = Metadata.getvalue(dataset, Metadata.OPERATION)
+        message = status
+
+        if not operations:
+            operations = []
+        elif did:
+            try:
+                operations.remove(did.name)
+            except ValueError:
+                pass
+
+        if enabled:
+            operations.extend([o.name for o in enabled if o.name not in operations])
+            operations.sort()
+            if not message:
+                message = "ok"
+
+        Metadata.setvalue(dataset, Metadata.OPERATION, operations)
```

Comment:
I like that. Aside from the minor fact that I need `set()` rather than `{}` which will always be interpreted as an empty *dict*, not an empty *set*. (Maybe if the interpreter itself actually cared about type hints, but this is not the language you're looking for ...)